### ISSUE TITLE
Performance issues/duplicate queries

### ISSUE THREAD

**Issue by MythThrazz** [easydigitaldownloads/easy-digital-downloads#6124]
Hello,

We've been investigating the performance of our page lately because we've started to have slowdowns, and while checking the backend we've found out that the Software Licensing backend page is doing a lot of duplicate queries (literally thousands).

Before the changes the Licenses page was loaded in nearly 7s and required over 3,6k queries. After few small tweaks we managed to get down to around 2s and around 100queries. Mainly by removing unnecessary updates and selects, which were repeated hundreds of times.

The changes are really small in most cases it's just saving the values statically within the function/method instead of calling the database multiple times.  

**easy-digital-downloads\includes\payments\class-edd-payment.php**

```
	private function setup_user_id() {
		$user_id  = $this->get_meta( '_edd_payment_user_id', true );
        /*
         * Cminds change
         */
        $old_user_id = $user_id;
		$customer = new EDD_Customer( $this->customer_id );

		// Make sure it exists, and that it matches that of the associted customer record
		if( empty( $user_id ) || ( ! empty( $customer->user_id ) && (int) $user_id !== (int) $customer->user_id ) ) {

			$user_id = $customer->user_id;

            /*
            * Cminds change
            */
            if($old_user_id != $user_id){
			// Backfill the user ID, or reset it to be correct in the event of data corruption
                $this->update_meta( '_edd_payment_user_id', $user_id );
            }

		}

		return $user_id;
	}
```

**easy-digital-downloads\includes\class-edd-customer.php**

         /**
	 * Given the customer data, let's set the variables
	 *
	 * @since  2.3
	 * @param  object $customer The Customer Object
	 * @return bool             If the setup was successful or not
	 */
	private function setup_customer( $customer ) {

		if ( ! is_object( $customer ) ) {
			return false;
		}

		foreach ( $customer as $key => $value ) {

            /*
            * Cminds change - removed the unnecessary notes loading
            */
            $this->$key = $value;

		}

		$this->emails   = (array) $this->get_meta( 'additional_email', false );
		$this->emails[] = $this->email;

		// Customer ID and email are the only things that are necessary, make sure they exist
		if ( ! empty( $this->id ) && ! empty( $this->email ) ) {
			return true;
		}

		return false;

	}

	/**
	 * Get things going
	 *
	 * @since 2.3
	 */
	public function __construct( $_id_or_email = false, $by_user_id = false ) {
        /*
         * Cminds change - added the static saving of the instances
         */
        static $_instances = array();

		$this->db = new EDD_DB_Customers;

		if ( false === $_id_or_email || ( is_numeric( $_id_or_email ) && (int) $_id_or_email !== absint( $_id_or_email ) ) ) {
			return false;
		}

		$by_user_id = is_bool( $by_user_id ) ? $by_user_id : false;

		if ( is_numeric( $_id_or_email ) ) {
			$field = $by_user_id ? 'user_id' : 'id';
		} else {
			$field = 'email';
		}

         /*
         * Cminds change - load the saved instance if found
         */
        if(empty($_instances[$field.$_id_or_email])){
            $customer = $this->db->get_customer_by( $field, $_id_or_email );
            $_instances[$field.$_id_or_email] = $customer;
        } else {
            $customer = $_instances[$field.$_id_or_email];
        }

		if ( empty( $customer ) || ! is_object( $customer ) ) {

			return false;
		}

		$this->setup_customer( $customer );

	}

**Comment by cklosowski** [comment]
@MythThrazz would you mind submitting this as a PR? That way it runs through our automated testing on Travis CI, and we can get a quick view into if it breaks anything else unexpectedly.

**Comment by chriscct7** [comment]
I've submitted prs #6146, #6145 and #6144 that cover the items above.

**Comment by chriscct7** [comment]
Also #6147

**Comment by chriscct7** [comment]
The combination of #6144, #6145, #6146 and #6147 (modified from the originally submitted proposal for speed and compatibility), when combined on `master` produces a 11 - 14% average speed improvement across all PHP, WP version, and MS (on/off) combinations on the EDD unit test system alone. When used on the software licensing page as well as the customer list table pages, there is a noticeable speed improvement, as EDD only does a fraction of the requests it currently does.

These 4 prs result in a significant performance improvement with no backwards compatibility concerns.

**Comment by chriscct7** [comment]
The combination also resulted in a significant reduction of queries in the unit test suite (9,490 queries eliminated, 11% of the total).

**Comment by cklosowski** [comment]
@chriscct7 ok so short story is that all your PRs that are open address this exact issue?

**Comment by chriscct7** [comment]
@cklosowski yes I split them into parts by what component they tackled to make it easier to test and run unit tests on. Combined they make an incredible speed improvement

**Comment by cklosowski** [comment]
:+1: thanks @chriscct7 I'll get it reviewed and merged.

**Comment by cklosowski** [comment]
I have to re-open this...this caused some major issues and fatal errors on not only Easy Digital Downloads, but on other customer sites.

We ran this for over 18 hours on our site before releasing, but didn't see the errors until nearly a day after. We'll have to re-address these issues.

**Comment by chriscct7** [comment]
I think a new ticket should be opened just for the one section (user id) causing issues. This issue is of 4 unrelated prs that touch completely different parts

**Comment by chriscct7** [comment]
Also what's the specific fatal error?
